packages/design/src/components/popup/RectPopup.tsx

```diff
@@ -0,0 +1,140 @@
+
+    onClickOther?: (e: MouseEvent) => void;
+}
+
+const calcHorizontalPopupPosition = (position: IAbsolutePosition, width: number, height: number, containerWidth: number, containerHeight: number): Partial<IAbsolutePosition> => {
+    const { left: startX, top: startY, right: endX, bottom: endY } = position;
+
+    const verticalStyle = ((startY + height) > containerHeight) ? { top: endY - height } : { top: startY };
+    const horizontalStyle = ((endX + width) > containerWidth) ? { left: startX - width } : { left: endX };
+
+    return {
+        ...verticalStyle,
+        ...horizontalStyle,
+    };
+};
+
+const calcVerticalPopupPosition = (position: IAbsolutePosition, width: number, height: number, containerWidth: number, containerHeight: number): Partial<IAbsolutePosition> => {
+    const { left: startX, top: startY, right: endX, bottom: endY } = position;
+
+    const verticalStyle = (endY + height) > containerHeight ? { top: startY - height } : { top: endY };
+    const horizontalStyle = (startX + width) > containerWidth ? { left: endX - width } : { left: startX };
+
+    return {
+        ...verticalStyle,
+        ...horizontalStyle,
+    };
+};
+
+export function RectPopup(props: IRectPopupProps) {
+    const { children, anchorRect, direction = 'vertical', mask, onMaskClick, onClickOther } = props;
+
+    const nodeRef = useRef(null);
+    const clickOtherFn = useRef(onClickOther);
+
+    clickOtherFn.current = onClickOther;
+    const [position, setPosition] = useState<Partial<IAbsolutePosition>>({
+        top: -9999,
+        left: -9999,
+    });
+
+    useEffect(() => {
+        const { clientWidth, clientHeight } = nodeRef.current!;
+        const { innerWidth, innerHeight } = window;
+        if (direction === 'horizontal') {
+            setPosition(
+                calcHorizontalPopupPosition(
+                    anchorRect,
+                    clientWidth,
+                    clientHeight,
+                    innerWidth,
+                    innerHeight
+                )
+            );
+        } else {
+            setPosition(
+                calcVerticalPopupPosition(
+                    anchorRect,
+                    clientWidth,
+                    clientHeight,
+                    innerWidth,
+                    innerHeight
+                )
+            );
+        }
+    }, [
+        anchorRect.left,
+        anchorRect.top,
+        anchorRect.bottom,
+        anchorRect.right,
+        direction,
+    ]);
+
+    useEffect(() => {
+        const handleClick = (e: MouseEvent) => {
+            clickOtherFn.current?.(e);
+        };
+        setTimeout(() => {
+            window.addEventListener('click', handleClick);
+        }, 100);
+        return () => {
+            window.removeEventListener('click', handleClick);
+        };
+    }, [clickOtherFn]);
+
+    return (
+        <>
+            {mask ? <div className={styles.popupMask} onClick={onMaskClick} /> : null}
```

Comment:
Will it prevent clicking events be dispatced to the canvas?

Reply:
cc @jikkai `ClickOutside` component may optimize user experience here.

Reply:
`mask` has beed removed